File: mobile/immobilier-sakane/lib/models/reception_whatsapp.dart
/// Groupe des types sans groupe connu.
const String groupeReceptionAutres = 'autres';

/// Un groupe de types de messages (réservations, nettoyage…).
class GroupeReceptionWhatsapp {
  final String code;
  final String libelle;

  const GroupeReceptionWhatsapp({required this.code, required this.libelle});

  factory GroupeReceptionWhatsapp.fromJson(Map<String, dynamic> json) {
    final code = (json['code'] ?? '').toString().trim();
    final libelle = (json['libelle'] ?? '').toString().trim();
    return GroupeReceptionWhatsapp(
      code: code,
      libelle: libelle.isEmpty ? code : libelle,
    );
  }
}

/// Un type de message WhatsApp que l'utilisateur peut recevoir ou couper.
/// Libelle et description viennent du serveur.
class TypeReceptionWhatsapp {
  final String code;
  final String libelle;
  final String description;
  final bool actif;

  /// Code du groupe ([groupeReceptionAutres] s'il manque).
  final String groupe;
  final String? groupeLibelle;

  const TypeReceptionWhatsapp({
    required this.code,
    required this.libelle,
    this.description = '',
    this.actif = true,
    this.groupe = groupeReceptionAutres,
    this.groupeLibelle,
  });

  factory TypeReceptionWhatsapp.fromJson(Map<String, dynamic> json) {
    final code = (json['code'] ?? '').toString();
    final groupe = (json['groupe'] ?? '').toString().trim();
    final groupeLibelle = (json['groupeLibelle'] ?? '').toString().trim();
    return TypeReceptionWhatsapp(
      code: code,
      libelle: (json['libelle'] ?? code).toString(),
      description: (json['description'] ?? '').toString(),
      actif: json['actif'] != false,
      groupe: groupe.isEmpty ? groupeReceptionAutres : groupe,
      groupeLibelle: groupeLibelle.isEmpty ? null : groupeLibelle,
    );
  }

  TypeReceptionWhatsapp copyWith({bool? actif}) {
    return TypeReceptionWhatsapp(
      code: code,
      libelle: libelle,
      description: description,
      actif: actif ?? this.actif,
      groupe: groupe,
      groupeLibelle: groupeLibelle,
    );
  }
}

/// Une section de la page : un groupe et ses types, dans l'ordre du serveur.
class SectionReceptionWhatsapp {
  final String code;
  final String libelle;
  final List<TypeReceptionWhatsapp> types;

  const SectionReceptionWhatsapp({
    required this.code,
    required this.libelle,
    required this.types,
  });

  int get nombreRecus => types.where((t) => t.actif).length;
  bool get toutRecu => types.every((t) => t.actif);
}

/// Ce qu'un utilisateur recoit comme notifications WhatsApp.
/// Les messages envoyes aux clients ne sont pas concernes.
class ReceptionWhatsapp {
  final int? managerId;
  final String? nom;
  final String? telephone;
  final bool actif;
  final List<GroupeReceptionWhatsapp> groupes;
  final List<TypeReceptionWhatsapp> types;

  const ReceptionWhatsapp({
    this.managerId,
    this.nom,
    this.telephone,
    this.actif = true,
    this.groupes = const [],
    this.types = const [],
  });

  factory ReceptionWhatsapp.fromJson(Map<String, dynamic> json) {
    final id = json['managerId'];
    return ReceptionWhatsapp(
      managerId: id is int ? id : int.tryParse('${id ?? ''}'),
      nom: json['nom']?.toString(),
      telephone: json['telephone']?.toString(),
      actif: json['actif'] != false,
      groupes: (json['groupes'] is List ? json['groupes'] as List : const [])
          .whereType<Map>()
          .map((e) => GroupeReceptionWhatsapp.fromJson(Map<String, dynamic>.from(e)))
          .where((g) => g.code.isNotEmpty)
          .toList(),
      types: (json['types'] is List ? json['types'] as List : const [])
          .whereType<Map>()
          .map((e) => TypeReceptionWhatsapp.fromJson(Map<String, dynamic>.from(e)))
          .where((t) => t.code.isNotEmpty)
          .toList(),
    );
  }

  ReceptionWhatsapp copyWith({
    bool? actif,
    List<TypeReceptionWhatsapp>? types,
  }) {
    return ReceptionWhatsapp(
      managerId: managerId,
      nom: nom,
      telephone: telephone,
      actif: actif ?? this.actif,
      groupes: groupes,
      types: types ?? this.types,
    );
  }

  /// Types regroupés par groupe : d'abord l'ordre des [groupes] du serveur,
  /// puis les groupes rencontrés dans les types, « Autres » en dernier.
  List<SectionReceptionWhatsapp> get sections {
    final parGroupe = <String, List<TypeReceptionWhatsapp>>{};
    for (final t in types) {
      parGroupe.putIfAbsent(t.groupe, () => []).add(t);
    }
    final libelles = <String, String>{
      for (final g in groupes) g.code: g.libelle,
    };
    for (final t in types) {
      if (t.groupeLibelle != null) libelles.putIfAbsent(t.groupe, () => t.groupeLibelle!);
    }
    final ordre = <String>[
      ...groupes.map((g) => g.code).where(parGroupe.containsKey),
      ...parGroupe.keys.where(
        (c) => c != groupeReceptionAutres && !groupes.any((g) => g.code == c),
      ),
    ];
    if (parGroupe.containsKey(groupeReceptionAutres) &&
        !ordre.contains(groupeReceptionAutres)) {
      ordre.add(groupeReceptionAutres);
    }
    return [
      for (final code in ordre)
        SectionReceptionWhatsapp(
          code: code,
          libelle: libelles[code] ??
              (code == groupeReceptionAutres ? 'Autres' : code),
          types: parGroupe[code]!,
        ),
    ];
  }

  bool get sansNumero => (telephone ?? '').trim().isEmpty;

  /// Codes des types coupes : c'est la liste complete qu'attend le serveur.
  List<String> get typesCoupes =>
      types.where((t) => !t.actif).map((t) => t.code).toList();

  int get nombreRecus => types.where((t) => t.actif).length;

  /// Resume court pour une liste : tout, rien, ou le nombre de types recus.
  String get resume {
    if (!actif || (types.isNotEmpty && nombreRecus == 0)) return 'Rien';
    if (nombreRecus == types.length) return 'Tout reçu';
    return '$nombreRecus type${nombreRecus > 1 ? 's' : ''} sur ${types.length}';
  }
}

File: backend/app/Services/ReceptionWhatsapp.php
<?php

namespace App\Services;

use App\Models\Manager;
use Illuminate\Support\Collection;

/**
 * Ce que chaque utilisateur accepte de recevoir sur WhatsApp, message par
 * message. Les envois aux clients et aux syndics ne sont pas concernes, ni
 * le code de reinitialisation du mot de passe.
 */
class ReceptionWhatsapp
{
    public const GROUPES = [
        "reservations" => "Réservations",
        "nettoyage"    => "Nettoyage",
        "charges"      => "Charges",
        "reclamations" => "Réclamations",
        "rappels"      => "Rappels",
    ];

    /** code => [groupe, libelle, description] */
    public const TYPES = [
        "reservation-ajoutee"     => ["reservations", "Réservation ajoutée", "Une nouvelle réservation est enregistrée."],
        "reservation-prolongee"   => ["reservations", "Réservation prolongée", "Un séjour est prolongé."],
        "reservation-raccourcie"  => ["reservations", "Réservation raccourcie", "Un séjour est raccourci."],
        "reservation-prix-modifie" => ["reservations", "Prix modifié", "Le prix d'une réservation est modifié."],
        "nettoyage-a-faire"       => ["nettoyage", "Appartement à nettoyer", "Un appartement est libéré et doit être nettoyé."],
        "nettoyage-commence"      => ["nettoyage", "Nettoyage commencé", "Un nettoyage démarre."],
        "nettoyage-termine"       => ["nettoyage", "Nettoyage terminé", "Un nettoyage est terminé."],
        "charge-ajoutee"          => ["charges", "Charge ajoutée", "Une nouvelle charge est saisie."],
        "charge-traitee"          => ["charges", "Charge traitée", "Une charge est validée."],
        "charge-programmee"       => ["charges", "Charge programmée", "Rappel d'une charge programmée."],
        "reclamation-ajoutee"     => ["reclamations", "Réclamation ajoutée", "Une nouvelle réclamation est signalée."],
        "reclamation-resolue"     => ["reclamations", "Réclamation résolue", "Une réclamation est traitée."],
        "rappels"                 => ["rappels", "Rappels d'arrivée", "Rappels des arrivées de vos clients."],
    ];

    /** Les codes d'un groupe, ou le code lui-meme. */
    private static function codes(string $type): array
    {
        if (isset(self::TYPES[$type])) {
            return [$type];
        }
        return array_keys(array_filter(self::TYPES, fn($t) => $t[0] === $type));
    }

    public static function typesCoupes(Manager $m): array
    {
        $coupes = $m->whatsapp_types_coupes;
        if (is_string($coupes)) {
            $coupes = json_decode($coupes, true);
        }
        $fins = [];
        foreach ((array) ($coupes ?? []) as $c) {
            foreach (static::codes((string) $c) as $f) $fins[] = $f;
        }
        return array_values(array_unique($fins));
    }

    public static function accepte(?Manager $m, string $type): bool
    {
        if (!$m) {
            return true;
        }
        if (!(bool) ($m->whatsapp_actif ?? true)) {
            return false;
        }
        $codes = static::codes($type);
        // Un groupe n'est refuse que si tous ses messages le sont.
        return $codes === [] || count(array_diff($codes, static::typesCoupes($m))) > 0;
    }

    /** Les 9 derniers chiffres : la meme personne, quel que soit le format saisi. */
    private static function cle(?string $tel): string
    {
        $chiffres = preg_replace('/\D/', '', (string) $tel);
        return strlen($chiffres) >= 9 ? substr($chiffres, -9) : $chiffres;
    }

    /** Retire des numeros ceux des utilisateurs qui refusent ce type de message. */
    public static function filtrer($telephones, string $type): Collection
    {
        $telephones = collect($telephones);
        if ($telephones->isEmpty()) {
            return $telephones;
        }

        $refus = Manager::query()
            ->where(fn($q) => $q->where("whatsapp_actif", false)->orWhereNotNull("whatsapp_types_coupes"))
            ->get(["id", "phone", "whatsapp_actif", "whatsapp_types_coupes"])
            ->filter(fn($m) => !static::accepte($m, $type))
            ->map(fn($m) => static::cle($m->phone))
            ->filter()
            ->flip();

        if ($refus->isEmpty()) {
            return $telephones->values();
        }

        return $telephones->reject(fn($t) => isset($refus[static::cle($t)]))->values();
    }

    public static function reglages(Manager $m): array
    {
        $coupes = static::typesCoupes($m);
        return [
            "managerId" => $m->id,
            "nom"       => trim(($m->first_name ?? "") . " " . ($m->last_name ?? "")),
            "telephone" => $m->phone,
            "actif"     => (bool) ($m->whatsapp_actif ?? true),
            "groupes"   => collect(self::GROUPES)->map(fn($libelle, $code) => ["code" => $code, "libelle" => $libelle])->values(),
            "types"     => collect(self::TYPES)->map(fn($t, $code) => [
                "code"        => $code,
                "groupe"      => $t[0],
                "groupeLibelle" => self::GROUPES[$t[0]] ?? $t[0],
                "libelle"     => $t[1],
                "description" => $t[2],
                "actif"       => !in_array($code, $coupes, true),
            ])->values(),
        ];
    }
}
